services/payments/tx_states.go

```diff
@@ -0,0 +1,174 @@
+package payments
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/google/uuid"
+)
+
+// TransactionState is an integer representing transaction status.
+type TransactionState string
+
+// TxStateMachine describes types with the appropriate methods to be Driven as a state machine
+const (
+	// Prepared represents a record that has been prepared for authorization.
+	Prepared TransactionState = "prepared"
+	// Authorized represents a record that has been authorized.
+	Authorized TransactionState = "authorized"
+	// Pending represents a record that is being or has been submitted to a processor.
+	Pending TransactionState = "pending"
+	// Paid represents a record that has entered a finalized success state with a processor.
+	Paid TransactionState = "paid"
+	// Failed represents a record that has failed processing permanently.
+	Failed TransactionState = "failed"
+)
+
+// Transitions represents the valid forward-transitions for each given state.
+var Transitions = map[TransactionState][]TransactionState{
+	Prepared:   {Authorized, Failed},
+	Authorized: {Pending, Failed},
+	Pending:    {Paid, Failed},
+	Paid:       {},
+	Failed:     {},
+}
+
+type baseStateMachine struct {
+	transaction *Transaction
+	service     *Service
```

Comment:
This feels a bit weird. Service is supposed to orchestrate all business operations. Data is supposed to implement algorithms on data. This breaks the contract – a state machine takes over and orchestrates service.

Can we achieve the same result by doing the orchestration in service?

The only place, as far as I can see, is in `populateInitialTransaction`. This seems like we need to find another way of passing the information it needs instead of making the state machine implementation dependant on the service.

Reply:
I've taken a pass at address this concern in [4a3c2a2](https://github.com/brave-intl/bat-go/pull/1821/commits/4a3c2a23e67842326e50f1d43d0271f213dec1b1). Take a look and let me know what you think.

Reply:
Thank you @Sneagan ! This addresses the concern with the service being available to the a state machine. 👍

Reply:
Resolving this.